src/betterproto/streams/__init__.py

```diff
@@ -0,0 +1,82 @@
+from .. import Message, encode_varint, decode_varint
+from io import BufferedReader, BufferedWriter
+from typing import Iterable, Optional
+
+
+class CodedOutputStream:
+    stream: BufferedWriter
+
+    def __init__(self, stream: BufferedWriter) -> None:
+        self.stream = stream
+
+    def write_varint(self, value: int) -> None:
+        """Write a single varint value to the stream."""
+        self.stream.write(encode_varint(value))
+
+    def write_message(self, message: Message) -> None:
+        """Write a serialized message to the stream."""
+        self.stream.write(bytes(message))
+
+    def write_messages(self, messages: Iterable[Message]) -> None:
+        """Write multiple messages to the stream."""
+        for message in messages:
+            # Write the size of the message to delimit it and enable parsing
+            self.write_varint(len(bytes(message)))
+            self.write_message(message)
+
+
+class CodedInputStream:
+    stream: BufferedReader
+
+    def __init__(self, stream: BufferedReader) -> None:
+        self.stream = stream
+
+    def read_varint(self) -> Optional[int]:
+        """Read a single varint from the stream. Returns None if stream is at EOF."""
+        byte = self.stream.read(1)
+        if byte == b"":
+            # EOF
+            return None
+
+        # Read extended varint bytes (if varint is longer than 1 byte)
+        value = 0
+        shift = 0
+        while int.from_bytes(byte, byteorder="little") >> 7 == 1:
+            # Add current byte to the value
+            value += (int.from_bytes(byte, byteorder="little") & 0b01111111) << (7 * shift)
+
+            # Read the next byte and check it is valid
+            byte = self.stream.read(1)
+            if byte == b"":
+                # EOF is invalid, as the full varint hasn't been read yet
+                raise ValueError("Stream ended part way through a varint")
+
+            shift += 1
+
+        # Add final byte of the varint to the value
+        value += (int.from_bytes(byte, byteorder="little") & 0b01111111) << (7 * shift)
+
+        return value
+
+    def read_message(self, size: int) -> Message:
+        """Read a message from the stream of a given size.
+        Returns empty bytes object if stream is at EOF."""
+        return self.stream.read(size)
+
+    def read_messages(self, limit=0) -> Iterable[Message]:
+        """Read messages from the stream given in the format
+        <varint message A length><message A><varint message B length>...
+        Will continue reading endlessly unless a message limit is set."""
+        size = self.read_varint()
+        count = 0
+        while size is not None:
+            yield self.read_message(size)
+            count += 1
+
+            # Get the size of the next message
+            if limit == 0 or count < limit:
+                size = self.read_varint()
+            else:
+                break
```

Comment:
Is it better to implement `Message.parse_delimited_from` by analogy with Java and use it here?